contributors/design-proposals/kubeadm-upgrades.md

```diff
@@ -0,0 +1,350 @@
+
+## Example user experience
+
+UX when upgrading a patch version:
+
+```console
+$ kubeadm upgrade plan
+[kubeadm upgrade plan output]
+```
+
+To upgrade the kubelets in the cluster, please upgrade them from the source you installed them from (most often using your package manager).
+
+```console
+$ kubeadm upgrade apply --version v1.8.3
+[kubeadm upgrade apply output]
+```
+
+## Upgrade planning
+
+In order for the user to know whether the cluster may be upgraded or not; the `kubeadm upgrade plan` command was written.
+The command:
+
+1. Checks if the cluster is healthy
+  - Does the API server `/healthz` endpoint return `ok`?
+  - Are all nodes in the cluster in the `Ready` state?
+  - (If Static Pods are used) Do all required Static Pod manifests exist on disk?
+  - (If cluster is self-hosted) Are all control plane DaemonSets healthy?
+1. Builds an internal configuration object to apply to the cluster
+  - The `--config` file takes precedence; otherwise the current configuration stored in a ConfigMap in the cluster is used.
+  - The `KubernetesVersion` field is set to the desired version.
+  - The built configuration object is optionally printed to STDOUT if `--print-config` is `true`.
+1. Information about what releases are available are downloaded from the Kubernetes Release & CI GCS buckets.
+  - The `dl.k8s.io/release/stable.txt` endpoint (and optionally `.../stable-1.X.txt`) is queried for the latest stable version to display.
+  - The API Server, kubeadm CLI, and all kubelets' versions are taken into account when computing the possible upgrades.
+  - The command may show zero to four upgrade possibilities.
+  - The information is nicely printed to the user.
+
+
+## Upgrade implementation
+
+On a high level, this is what `kubeadm upgrade apply [version]` does:
+
+1. Checks if the cluster is healthy
+  - Does the API server `/healthz` endpoint return `ok`?
+  - Are all nodes in the cluster in the `Ready` state?
+  - (If Static Pods are used) Do all required Static Pod manifests exist on disk?
+  - (If cluster is self-hosted) Are all control plane DaemonSets healthy?
+1. Builds an internal configuration object to apply to the cluster
+  - The `--config` file takes precedence; otherwise the current configuration stored in a ConfigMap in the cluster is used.
+  - The `KubernetesVersion` field is set to the desired version.
+  - The built configuration object is optionally printed to STDOUT if `--print-config` is `true`.
+1. Enforces the version policies highlighted above in [Pre-upgrade policies](#pre-upgrade-policies)
+1. Asks the user for a `[Y|n]` answer whether to proceed with the upgrade or not (if the session is interactive)
+1. Performs a method of ensuring that all necessary control plane images are available locally to the master(s), aka. "pre-pulling"
+  - Due to lack of an API to the kubelet that would make it possible to say "make sure this image X is available locally"; we have to
```

Comment:
Have we filed a feature request against sig-node to add such an API?

Reply:
somewhere there is such an issue..., but we haven't escalated it more broadly IIRC